File: backend/services/post-ktor-mongo/src/main/kotlin/me/jason5lee/post_ktor_mongo_fdm/list_posts/impl.kt
package me.jason5lee.post_ktor_mongo.list_posts

import com.mongodb.client.model.Projections
import kotlinx.coroutines.flow.toList
import me.jason5lee.post_ktor_mongo.common.*
import me.jason5lee.post_ktor_mongo.common.utils.Db
import me.jason5lee.post_ktor_mongo.common.utils.Deps
import org.bson.Document
import org.bson.types.ObjectId

class WorkflowImpl(private val deps: Deps) : Workflow(), ErrorsImpl {
    override suspend fun run(input: Query): Output {
        val filter = Document()
        val creatorMap: MutableMap<ObjectId, UserName> = mutableMapOf()
        if (input.creator != null) {
            val creatorOid = Db.toObjectIdOrNull(input.creator.value) ?: throw creatorNotFound()
            val creatorUserDoc = Db.findById(
                db = deps.mongoDb,
                collection = Db.users,
                id = creatorOid,
                projection = Projections.include("name"),
            ) ?: throw creatorNotFound()
            creatorMap[creatorOid] = UserName(creatorUserDoc.get("name"))
            filter["creator"] = creatorOid
        }

        val total = Db.countExact(
            db = deps.mongoDb,
            collection = Db.posts,
            filter = filter,
        )
        val posts = Db.find<ObjectId>(
            db = deps.mongoDb,
            collection = Db.posts,
            configCollection = {
                this.find(filter)
                    .sort(Document("creationTime", -1).append("_id", -1))
                    .skip((input.page.value - 1) * input.pageSize.value)
                    .limit(input.pageSize.value)
            }
        ).toList()

        if (posts.isEmpty()) {
            return Output(
                total = total,
                posts = emptyList(),
            )
        }

        if (creatorMap.isEmpty()) {
            val creatorIds = posts.mapTo(HashSet()) { it.get<ObjectId>("creator") }
            val creatorDocs = Db.find<ObjectId>(
                db = deps.mongoDb,
                collection = Db.users,
                configCollection = {
                    this.find(Document("_id", Document("\$in", creatorIds)))
                        .projection(Projections.include("name", "_id"))
                }
            ).toList()
            for (creatorDoc in creatorDocs) {
                val creatorId = creatorDoc.get<ObjectId>("_id")
                creatorMap[creatorId] = UserName(creatorDoc.get("name"))
            }
        }

        return Output(
            total = total,
            posts = posts.map { postDoc ->
                val creatorId = postDoc.get<ObjectId>("creator")
                val creatorName = creatorMap[creatorId]
                    ?: throw Exception("Invalid query result from `${Db.posts}[_id = ${postDoc.id}].creator`, not found in `${Db.users}`")

                Post(
                    id = PostId(Db.formatId(postDoc.id)),
                    title = Title(postDoc.get("title")),
                    creator = Creator(
                        id = UserId(Db.formatId(creatorId)),
                        name = creatorName,
                    ),
                    creationTime = Time(postDoc.get("creationTime")),
                )
            }
        )
    }
}

File: backend/services/post-ts-koa-mongo/src/list-posts/impl.ts
import { Deps } from "../common/utils";
import { Output, Query, Workflow } from ".";
import * as db from "../common/utils/db";
import { WithId, ObjectId } from "mongodb";
import { PostId, Time, Title, UserId, UserName } from "../common";
import { errors } from "./api";
import * as runtypes from "runtypes";

export class WorkflowImpl implements Workflow {
    private static readonly expectedPost = runtypes.Record({
        title: runtypes.String,
        creator: runtypes.InstanceOf(ObjectId),
        creationTime: runtypes.Number,
    });
    private static readonly userHasName = runtypes.Record({
        name: runtypes.String,
    });
    async run(query: Query): Promise<Output> {
        let filter: object = {};
        if (query.creator !== undefined) {
            const creatorOid = db.tryParseId(query.creator);
            if (creatorOid === undefined) {
                throw this.errors.creatorNotFound();
            }
            const user = await this.deps.mongoDb.collection(db.users).findOne({ _id: creatorOid });
            if (user === null) {
                throw this.errors.creatorNotFound();
            }
            filter = { creator: creatorOid };
        }

        const total = await this.deps.mongoDb
            .collection(db.posts)
            .countDocuments(filter);

        const pagePosts: WithId<unknown>[] = await this.deps.mongoDb
            .collection(db.posts)
            .find(filter)
            .sort({ creationTime: -1, _id: -1 })
            .skip((query.page - 1) * query.pageSize)
            .limit(query.pageSize)
            .toArray();
        if (pagePosts.length === 0) {
            return { total, posts: [] };
        }
        db.validateArray(WorkflowImpl.expectedPost, db.posts, pagePosts);
        const creatorIds = pagePosts.map(post => post.creator);
        const creators: WithId<unknown>[] = await this.deps.mongoDb.collection(db.users).find({ _id: { $in: creatorIds } }).toArray();
        db.validateArray(WorkflowImpl.userHasName, db.users, creators);

        const creatorMap = new Map(creators.map(creator => [creator._id.toHexString(), creator]));
        return {
            total,
            posts: pagePosts.map(post => {
                const creator = creatorMap.get(post.creator.toHexString());
                if (creator === undefined) {
                    throw new Error(`The creator of post ${post._id}, ${post.creator}, not found`);
                }

                return {
                    id: db.formatId(post._id) as PostId,
                    title: post.title as Title,
                    creator: {
                        id: db.formatId(post.creator) as UserId,
                        name: creator.name as UserName,
                    },
                    creationTime: { utc: post.creationTime } as Time,
                };
            })
        };
    }

    readonly errors = errors;
    constructor(private readonly deps: Deps) { }
}
